musica/musica_ccpp.F90

```diff
@@ -9,52 +10,116 @@ module musica_ccpp
-    real(kind_phys),                   intent(inout) :: constituents(:,:,:)  ! kg kg-1
+    use ccpp_kinds,                only: kind_phys
+    use iso_c_binding,             only: c_double
+    real(kind_phys),                   intent(inout) :: time_step            ! s
+    real(kind_phys), target,           intent(inout) :: temperature(:,:)     ! K
+    real(kind_phys), target,           intent(inout) :: pressure(:,:)        ! Pa
+    real(kind_phys), target,           intent(inout) :: dry_air_density(:,:) ! kg m-3
+    type(ccpp_constituent_prop_ptr_t), intent(inout) :: constituent_props(:)
+    real(kind_phys), target,           intent(inout) :: constituents(:,:,:)  ! kg kg-1
+    real(kind_phys), target,           intent(inout) :: rate_params(:,:,:)
+    real(kind_phys), target,           intent(in)    :: height(:,:)          ! km
     character(len=512),                intent(out)   :: errmsg
     integer,                           intent(out)   :: errcode
 
-    call micm_run(time_step, temperature, pressure, dry_air_density, constituent_props, &
-                  constituents, errmsg, errcode)
+    ! local variables
+    real(c_double), target, dimension(size(temperature, dim=1)     &
+                                    * size(temperature, dim=2))      :: m_temperature
+    real(c_double), target, dimension(size(pressure, dim=1)        &
+                                    * size(pressure, dim=2))         :: m_pressure
+    real(c_double), target, dimension(size(dry_air_density, dim=1) &
+                                    * size(dry_air_density, dim=2))  :: m_dry_air_density
+    real(c_double), target, dimension(size(constituents, dim=1)    &
+                                    * size(constituents, dim=2)    & 
+                                    * size(constituents, dim=3))     :: m_constituents ! mol m-3
+    real(c_double), target, dimension(size(rate_params, dim=1)     &
+                                    * size(rate_params, dim=2)     & 
+                                    * size(rate_params, dim=3))      :: m_rate_params
+    real(kind_phys), target, dimension(size(constituents, dim=3))    :: molar_mass_arr ! kg mol-1
+    integer :: i_elem
+
+    call tuvx_run(height, temperature, dry_air_density, errmsg, errcode)
+
+    ! Get the molar_mass that is set in the call to instantiate()
+    do i_elem = 1, size(molar_mass_arr)
+      call constituent_props(i_elem)%molar_mass(molar_mass_arr(i_elem), errcode, errmsg)
+      if (errcode /= 0) then
+        errmsg = "[MUSICA Error] Unable to get molar mass."
+        return
+      end if
+    end do
+
+    ! TODO(jiwon) Check molar mass is non zero as it becomes a denominator for unit converison
+    ! this code needs to go when ccpp framework does the check
+    do i_elem = 1, size(molar_mass_arr)
+      if (molar_mass_arr(i_elem) == 0) then
+        errcode = 1
+        errmsg = "[MUSICA Error] Molar mass must be a non zero value."
```

Comment:
Yeah that sound reasonable, done. This is temporary code knowing that the framework will eventually do the check. Do we have any updates on this?

Reply:
My guess is that it will probably still be a while (few months?) until this error-checking is enabled in the framework, unless NOAA swoops in and does it for us.  Of course if you find that it is starting to cause problems on your end then we can always see about moving it up in terms of priority.